bin/gcs_cli.py

```diff
@@ -119,6 +120,46 @@ def list_objects(bucket_name, details):
         click.echo("No objects in bucket.")
 
 
+@gcs_group.command("upload")
+@click.argument("source")
+@click.argument("destination")
+def upload(source, destination):
+    """Upload files to a bucket"""
+
+    client = get_client()
+
+    # remove protocol from destination if present
+    destination = destination.split("://", 1)[-1]
+    bucket_name, _, prefix = destination.partition("/")
+
+    try:
+        bucket = client.get_bucket(bucket_name)
+    except NotFound:
+        click.error(f"GCS bucket {bucket_name!r} does not exist.")
+        return
+
+    source_path = Path(source)
+    if not source_path.exists():
+        click.error("local path {source!r} does not exist.")
+    if source_path.is_dir():
+        prefix_path = Path(prefix)
+        sources = [p for p in source_path.rglob("*") if not p.is_dir()]
+    else:
+        sources = [source_path]
+    if not sources:
+        click.echo("No files in directory {source!r}.")
+        return
+    for path in sources:
+        if path == source_path:
+            key_path = prefix_path
+        else:
+            key_path = prefix_path / path.relative_to(source_path)
+        key = "/".join(key_path.parts)
```

Comment:
tl;dr I should make `prefix_path` a `PurePosixPath` and change this to `str(key_path)`

```suggestion
        key = str(key_path)
```

explanation:

1. `key_path` should not use absolute, because it is not a local-filesystem path and should not be prefixed with the current working directory.
2. this is platform agnostic as-is, and your change would it non-functional on windows:

   - `pathlib.Path` returns a platform-sensitive obect, `WindowsPath` or `PosixPath`, and because the code doesn't specify a platform, all of the path objects here are platform-sensitive.
   - `key` must be `/` separated regardless of the current platform, and `"/".join(key_path.parts)` forcibly produces a `/` separated result from either a `WindowsPath` or `PosixPath`.
   - `prefix_path` should be instantiated as a `PurePosixPath`, because it must always be `/` separated even on windows, because it's the path part of a gcs URI (the `a/b` part of `gs://bucket/a/b`)
      - doing that on windows makes the line above do `PurePosixPath / WindowsPath` which produces a `PurePosixPath` and converts all of the windows `\`s to `/`s as necessary, and allows this line to be `key = str(key_path)`